HTCONDOR-842 File Transfer plugin errors should always generate a meaningful hold message, subcode

markcoatsworth opened the pull request:
Our File Transfer code doesn't provide meaningful error reporting when plugins exit unexpectedly. We want to improve this in the following ways:

* When a plugin exits unexpectedly without producting any classad output, indicate this in the job ad HoldReason message (instead of leaving it blank)
* If a plugin produces invalid classad output, indicate this in the job ad HoldReason
* If a plugin produces valid classad output but does not include required attributes (TransferSuccess and TransferError), indicate this in the job ad HoldReason
* Standardized plugin exit codes (currently: 0 = success, 1 = false, 2 = invalid credentials) and update existing plugins to actually use these codes
* If a plugin fails, set the job ad HoldReasonSubcode to the exit code left-shifted by 8 (this left-shift allows us to differentiate between plugin failures and regular cedar failures)

# HTCondor Pull Request Checklist for internal reviewers

- [x] Verify that (GitHub thinks) the merge is clean. If it isn't, and you're confident you can resolve the conflicts, do so. Otherwise, send it back to the original developer.
- [ ] Verify that the related Jira ticket exists and has a target version number and that it is correct.
- [x] Verify that the Jira ticket is in review status and is assigned to the reviewer.
- [x] Verify that the Jira ticket (HTCONDOR-xxx) is mentioned at the beginning of the title. Edit it, if not
- [ ] Verify that the branch destination of the PR matches the target version of the ticket
- [ ] Check for correctness of change
- [x] Check for regression test(s) of new features and bugfixes (if the feature doesn't require root)
- [ ] Check for documentation, if needed
- [x] Check for version history, if needed
- [x] Check BaTLab dashboard for successful build (https://batlab.chtc.wisc.edu/results/workspace.php) and test for either the PR or a workspace build by the developer that has the Jira ticket as a comment.
- [x] Check that commit message references the Jira ticket (HTCONDOR-xxx)

## After the above
- Hit the merge button if the pull request is approved and it is not a security patch (security changes require 2 additional reviews)
- If the pull request is approved, take the ticket out of review state
- Assign JIRA Ticket back to the developer

Todd-L-Miller commented on src/condor_utils/file_transfer.cpp:
```diff
@@ -5809,13 +5828,21 @@ FileTransfer::InvokeFileTransferPlugin(CondorError &e, const char* source, const
 
+	if (e.getFullText().empty()) {
+		e.pushf("FILETRANSFER", 1, "File transfer plugin %s exited unexpectedly "
+			"without producing an error message\n", plugin.c_str());
```
In the case that `result == TransferPluginResult::Success`, in which case we should resonably expect `e.getFullText()` to be empty, should we really be pushing an error message?

markcoatsworth replied:
Thanks, this was a misplaced check that I moved somewhere else and then forgot to remove. I'll push an update.

Todd-L-Miller reviewed (commented):
* The ticket is currently targeting 9.0.x, but the PR is targeting master and the version history is for 9.6.0.  I assume this means the ticket needs to updated?
* Do we have documentation for file-transfer plugin authors anywhere?  If so, we should probably document the new-standard exit codes there.
* Using constants in Python scripts would make me happier, even if they're just defined in the plug-in with a comment pointing to the docs in the previous point.

markcoatsworth commented:
I've addressed the changes requested. A few quick comments:

- This change is now targeting master for the v9.6.0 release.
- I've updated the documentation with standard error codes, also included these in the example plugin using constants as requested.
- The `e.getFullText().empty()` check you questioned above was something that I forgot to remove, it's gone now.

Todd-L-Miller reviewed (approved):
Changes look good, thanks.
